Gazebo: add sailing catarmaran model

srmainwaring opened the pull request:
Sailing Catamaran designed by Andy Little ([@kwikius](https://github.com/kwikius))

![catarmaran_rigged_3](https://github.com/ArduPilot/SITL_Models/assets/24916364/f0a65fc0-f25b-43ea-8690-f7eb979ed455)

## Usage

Gazebo and the plugins should be installed as per the [ArduPilot Gazebo Plugin](https://github.com/ArduPilot/ardupilot_gazebo) instructions.

Update the `GZ_SIM_RESOURCE_PATH` to include these models:

```bash
export GZ_SIM_RESOURCE_PATH=$GZ_SIM_RESOURCE_PATH:\
$HOME/SITL_Models/Gazebo/models:\
$HOME/SITL_Models/Gazebo/worlds
```

#### Waves model

The model is configured to work with the hydrodynamics plugin from the [wave sim](https://github.com/srmainwaring/asv_wave_sim) package. Follow the package instructions to set up the marine simulation.

#### Sail, foil, wind sensor and wind plugins

A number of additional plugins are required to support sailing vessels. These can be found at [asv_sim](https://github.com/srmainwaring/asv_sim). Installation instructions are in the repository README.

#### Run Gazebo

```bash
gz sim -v4 -r catamaran_waves.sdf
```
#### Run ArduPilot SITL

```bash
sim_vehicle.py -D -v Rover --model JSON --add-param-file=$HOME/SITL_Models/Gazebo/config/catamaran.param --console --map --custom-location='51.566151,-4.034345,10.0,-135'
```

#### Wave and wind settings

To run the simulation in calm seas modify the wave settings in `asv_wave_sim/gz-waves-models/world_models/waves/model.sdf`. There are two occurrences of the parameter block:

```xml
<wave>
  <!-- `fft` wave parameters -->
  <algorithm>fft</algorithm>
  <tile_size>256</tile_size>
  <cell_count>256</cell_count>
  <wind_speed>5.0</wind_speed>
  <wind_angle_deg>135</wind_angle_deg>
  <steepness>2</steepness>
</wave>
``` 
Update both blocks when making changes (one is for the physics, one for the visuals).

- Set `<wind_speed>` to 0 for calm seas.
- Set  `<cell_count>` to 4 to reduce the compute load if waves are not required.

The `catamaran_waves.sdf` world file is configured to load a wind plugin that responds to service requests to set the wind. For ease of gyro and accelerometer calibration the wind is initially set to zero. Once the vehicle is initialised and armed the wind can be set using:

```bash
gz topic -t /wind -m  gz.msgs.Vector3d -p "x:{wind_vel_x}, y:{wind_vel_y}, z:0"
```

where `{wind_vel_x}` and `{wind_vel_y}` should be replaced by the desired components of the wind velocity in `m/s`. A good choice is `"x:0, y:5, z:0"` to get the boat moving, then once out of irons switch to `"x:5.5, y:5.5, z:0"`

The [sail module](https://ardupilot.org/mavproxy/docs/modules/sail.html) in MAVProxy is useful for determining the current true and apparent wind direction.

kwikius commented:
Looks great.  My issue is that it needs a high end PC to work with the waves. That is a pity because you will lose a lot of potential users, who after getting through the installation hurdles, will then find that it locks up their PC. Maybe add another "world" where there are no waves? ( I will try to do that but will take a while!)
Also wonder about chopping wind vane ( and its spar) from the rig and attaching to the front of the pod, which may be a more normal ardupilot config. Currently anyway it doesnt seem to function. I do prefer a wind vane  ( and perhaps a cup anemometer) to the ultrasonic one because the visual is dynamic.

Did you increase the rig size or is that just the view. Bigger rig will of course pitchpole more easily. We could just make the hulls closer together to get it flying a hull, then less likely to pitchpole. Ideally width between hulls would be adjustable as well.
Anyway it looks fantastic in the gazebo sim

kwikius commented:
Other thing I was wondering about was making a simpler version of the hukks (and everything else) with less mesh points. Could use 1/2 the number of hull stations makbe and just simple flat plae foils, and remove complexity from the rig. Could also just use a normal mast rather than the z bend one.  Not sure if that would help to speed up the sim on slow pc?

srmainwaring commented:
- Gazebo running with ogre2 rendering is a demanding application in any case, and the waves simulation adds more to that still. I'll look into providing a version that does not simulate waves, just the hydrodynamics for flat water.
- The wind vane is rigged to move, but I am not using it as a sensor (and it does not have a lift plugin attached, so just spins about). The green transparent cylinder denotes the position of the anemometer - it's fixed in at the location of the wind vane when the rig is centred - so the readings passed to SITL are correct.
- Rig is as per provided STL file. I've converted to collada (DAE) for visuals, and scaled everything by 1E-3 as Gazebo expects units of `m` not `mm`.
- The hull collision meshes are pretty good already, and I've simplified the foils. If you look in the `./meshes` subfolder you'll see separate STL files for the collisions. The vertex count for each is:
  - hull_left_collision.stl: vertices: 121, triangles: 238
  - hull_right_collision.stl: vertices: 123, triangles: 238
  - fin: vertices: 28, triangles: 52
  - rudder: vertices: 28, triangles: 52
  - wind sensor: vertices: 63, triangles: 122
- The hull collisions were obtained by removing the platform and cross beams then reconstructing the affected faces. The foils and wind sensor were obtained by applying a decimation modifier to the original meshes.
- Changing the rig mesh won't affect performance significantly. Gazebo allows you to specify the meshes for visuals separately from those for collisions (hit box in the game world). Once loaded the visuals are handled by the GPU, so provided memory is not a problem they don't significantly affect performance. The collision meshes do, as too higher count affects the contacts calculation and in the case of hydrodynamics affects the physics. I don't think in this case the triangle count is the bottle-neck - most likely the wave physics update and render is the cause.
